### PR TITLE
Add ContainerRecreateRequest API definition

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

### Ⅰ. Describe what this PR does
Add ContainerRecreateRequest API definition

Reference to proposal [ContainerRecreateRequest](https://github.com/FillZpp/kruise/blob/proposal-for-container-recreate-request/docs/proposals/20210316-containerrecreaterequest.md).

### CODE REVIEW

File: apis/apps/v1alpha1/containerrecreaterequest_types.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// ContainerRecreateRequestContainer defines the container that need to recreate.
+type ContainerRecreateRequestContainer struct {
+	// Name of the container that need to recreate.
+	// It must be existing in the real Pod containers.
+	Name string `json:"name"`
+	// SkipPreStop indicates whether to skip the preStop hook during recreating this container.
+	SkipPreStop bool `json:"skipPreStop,omitempty"`
```
Comment (furykerry) [inline apis/apps/v1alpha1/containerrecreaterequest_types.go:65]: is it necessary to include PreStop, ports and StatusContext fields? is it ok to  just use the config in pod spec

File: apis/apps/v1alpha1/containerrecreaterequest_types.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+// ContainerRecreateRequestStrategy contains the strategies for containers recreation.
+type ContainerRecreateRequestStrategy struct {
+	// FailurePolicy decides whether to continue if one container fails to recreate
+	FailurePolicy ContainerRecreateRequestFailurePolicyType `json:"failurePolicy,omitempty"`
+	// The priority of grace period usage:
+	// stopGracePeriodSeconds > pod.Spec.TerminationGracePeriodSeconds > 3s
+	StopGracePeriodSeconds *int64 `json:"stopGracePeriodSeconds,omitempty"`
+	// ContinueByReady indicates whether to recreate the next container only if the previous one has been recreated and ready
+	ContinueByReady bool `json:"continueByReady,omitempty"`
```
Comment (furykerry) [inline apis/apps/v1alpha1/containerrecreaterequest_types.go:96]: please consider release ContinueByReady and ReadyProtection features in future patches

File: apis/apps/v1alpha1/containerrecreaterequest_types.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+	// TTLSecondsAfterFinished is the TTL duration after this ContainerRecreateRequest has completed.
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// ContainerRecreateRequestContainer defines the container that need to recreate.
+type ContainerRecreateRequestContainer struct {
+	// Name of the container that need to recreate.
+	// It must be existing in the real Pod containers.
+	Name string `json:"name"`
```
Comment (Fei-Guo) [inline apis/apps/v1alpha1/containerrecreaterequest_types.go:63]: I am a little confused here. Is the name the "Names" field shown in the `docker ps` cmd? or the container name shown in the Pod spec? 

If it is the latter, how can people specify the Pod identity in this API?

Comment (FillZpp) [reply apis/apps/v1alpha1/containerrecreaterequest_types.go:62]: Have added 'spec.podName=xxx'. But the webhook will still sync it into labels in order to get or list CRR using pod name as label selector.

**Review by furykerry — COMMENTED** [review]
LGTM

File: apis/apps/v1alpha1/containerrecreaterequest_types.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	// TTLSecondsAfterFinished is the TTL duration after this ContainerRecreateRequest has completed.
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// ContainerRecreateRequestContainer defines the container that need to recreate.
+type ContainerRecreateRequestContainer struct {
+	// Name of the container that need to recreate.
+	// It must be existing in the real pod.Spec.Containers.
+	Name string `json:"name"`
+	// StatusContext is synced from the real container in Pod spec during this ContainerRecreateRequest creating.
```
Comment (Fei-Guo) [inline apis/apps/v1alpha1/containerrecreaterequest_types.go:63]: Fix comment pls.

File: apis/apps/v1alpha1/containerrecreaterequest_types.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	// Populated by the system.
+	// Read-only.
+	PreStop *v1.Handler `json:"preStop,omitempty"`
+	// StatusContext is synced from the real container in Pod spec during this ContainerRecreateRequest creating.
```
Comment (Fei-Guo) [inline apis/apps/v1alpha1/containerrecreaterequest_types.go:67]: Fix comment.

File: apis/apps/v1alpha1/containerrecreaterequest_types.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	// It is only synchronized during a ContainerRecreateRequest in Recreating phase.
+	ContainerRecreateRequestSyncContainerStatusesKey = "crr.apps.kruise.io/sync-container-statuses"
+	// ContainerRecreateRequestUnreadyAcquiredKey indicates the Pod has been forced to not-ready.
+	// It is required if the unreadyGracePeriodSeconds is set in ContainerRecreateRequests.
+	ContainerRecreateRequestUnreadyAcquiredKey = "crr.apps.kruise.io/unready-acquired"
+)
+
+// ContainerRecreateRequestSpec defines the desired state of ContainerRecreateRequest
+type ContainerRecreateRequestSpec struct {
+	// PodName is name of the Pod to recreate.
```
Comment (Fei-Guo) [inline apis/apps/v1alpha1/containerrecreaterequest_types.go:46]: PodName is name of the Pod that owns the recreated containers.

**Review by Fei-Guo — COMMENTED** [review]
Mostly LG. Nits

**Review by furykerry — APPROVED** [review]

**Pull request merged**